velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +67,355 @@ void generateJsonTyped(
     int row,
     std::string& result) {
   auto value = input.valueAt(row);
-  folly::toAppend<std::string, T>(value, &result);
-}
 
-template <>
-void generateJsonTyped<StringView>(
-    const SimpleVector<StringView>& input,
-    int row,
-    std::string& result) {
-  auto value = input.valueAt(row);
-  folly::json::escapeString(value, result, folly::json::serialization_opts{});
+  if constexpr (std::is_same_v<T, StringView>) {
+    folly::json::escapeString(value, result, folly::json::serialization_opts{});
+  } else if constexpr (std::is_same_v<T, bool>) {
+    result.append(value ? "true" : "false");
+  } else if constexpr (
+      std::is_same_v<T, Date> || std::is_same_v<T, Timestamp>) {
+    result.append(std::to_string(value));
+  } else {
+    folly::toAppend<std::string, T>(value, &result);
+  }
 }
 
-template <TypeKind kind>
-void castPrimitiveToJson(
+// Casts primitive-type input vectors to Json type. If throwIfNull is true,
+// raises an exception when the input vector contains a null at any of selected
+// rows.
+template <
+    TypeKind kind,
+    bool throwIfNull = false,
+    typename std::enable_if<TypeTraits<kind>::isPrimitiveType, int>::type = 0>
+void castToJson(
     const BaseVector& input,
+    exec::EvalCtx* /*context*/,
     const SelectivityVector& rows,
     FlatVector<StringView>& flatResult) {
   using T = typename TypeTraits<kind>::NativeType;
 
+  // input is guaranteed to be in flat encoding when passed in.
```

Comment:
`SimpleVector` does not imply flat encoding. Nor is flat encoding guaranteed here; the DecodedVector::base() can return a constant.

Why not use a `DecodedVector` here? If you move the handling of encodings to castToJson, you can avoid repeating it when handling the child vectors for maps, arrays, and structs.

Reply:
That's true for the first call to `castTo` and made sense before complex type support was added, but recursive calls to `castTo` end up having to re-implement the logic to operate over the base vector. IMO overall it would be simpler to centralize this handling in `castTo` now.

Reply:
I haven't addressed this comment in the current version. I thought about it and realized that if I move the code of decoding the input vector, translating rows to base rows, and wrapping the result vector into JsonCastOperator, they will need to be inside `castToTyped` because this is the recursive function. So I'll still have these logics in each version of `castToTyped` (for primitive/array/map/row input vector types). Am I understanding it correctly?